### PR TITLE
WIP add library work on bugs

### PR DESCRIPTION
**Pull request by PavelBoPavel**

### CODE REVIEW

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+           if (authorToBookNamesMap.get(book.getAuthorBook()).size() == 0) {
+               authorToBookNamesMap.remove(book.getAuthorBook());
+           }
+           return true;
+       }catch (NullPointerException e){
+           return false;
+       }
+    }
+
+    public Book findByBook(String bookName){
```
Comment (KaydunovDenis) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:38]: bookName or name Book? See you class Book.
I think these method need rename:
findByBook->findByNameBook

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        for (Map.Entry<String, ArrayList<Book>> authorIsBooks : authorToBookNamesMap.entrySet()){
+            ArrayList<Book> books = authorIsBooks.getValue();
```
Comment (KaydunovDenis) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:40]: Best practic use type "var"

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public void addBook(Book book){
+        int temp = 0;
+        for (Map.Entry<String, ArrayList<Book>> authorBooks : authorToBookNamesMap.entrySet()){
+            if (authorBooks.getKey().equals(book.getAuthorBook())){
+                authorBooks.getValue().add(book);
+                temp++;
+            }
+        }
+        if (temp ==0 || authorToBookNamesMap.size() == 0){
+            ArrayList<Book> books = new ArrayList<>();
```
Comment (KaydunovDenis) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:20]: Best practic use type "var"

**Review by KaydunovDenis — CHANGES_REQUESTED** [review]

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java (hunk 2 above)
Comment (PavelBoPavel) [reply homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:40]: в данном случае не согласен. по значению справа от знака равно явно не понятен тип

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java (hunk 3 above)
Comment (PavelBoPavel) [reply homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:20]: согласен

**Review by VasiliZas — APPROVED** [review]

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            books.add(book);
+            authorToBookNamesMap.put(book.getAuthorBook(), books);
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:22]:
```
if (authorToBookNamesMap.containsKey(book.getAuthorBook())) {
    authorBooks.getValue().add(book);
} else {
    authorToBookNamesMap.put(book.getAuthorBook(), List.of(books));
}
```

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public boolean removeBook(Book book){
+       try {
+           authorToBookNamesMap.get(book.getAuthorBook()).remove(book);
+           if (authorToBookNamesMap.get(book.getAuthorBook()).size() == 0) {
+               authorToBookNamesMap.remove(book.getAuthorBook());
+           }
+           return true;
+       }catch (NullPointerException e){
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:33]: Best practices to write code the way that prevent NPE to be thrown, not catching it.

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        if (temp ==0 || authorToBookNamesMap.size() == 0){
+            ArrayList<Book> books = new ArrayList<>();
+            books.add(book);
+            authorToBookNamesMap.put(book.getAuthorBook(), books);
+        }
+    }
+
+    public boolean removeBook(Book book){
+       try {
+           authorToBookNamesMap.get(book.getAuthorBook()).remove(book);
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:28]: At this point you should check if List under key is present.

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+           if (authorToBookNamesMap.get(book.getAuthorBook()).size() == 0) {
+               authorToBookNamesMap.remove(book.getAuthorBook());
+           }
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:31]: In practice, no need for this.

File: homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    public Book findByBook(String bookName){
+        for (Map.Entry<String, ArrayList<Book>> authorIsBooks : authorToBookNamesMap.entrySet()){
+            ArrayList<Book> books = authorIsBooks.getValue();
+            for (int i = 0; i < books.size(); i++){
+               if(books.get(i).getNameBook().equals(bookName))
+                   return books.get(i);
+            }
+        }
+        return null;
+    }
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/main/java/com/tutrit/java/quickstart/Library.java:47]:
```
if (authorToBookNamesMap.containsKey(book.getAuthorBook()) {
    List<Books> books = authorToBookNamesMap.get(book.getAuthorBook());
    for (Book book : books) {
        if(book.getNameBook().equals(bookName)) {
            return book;
        }
    }
}
return null;
```

File: homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        bookPuskin3 = new Book("Pushkin", "starushka");
+        bookChehov = new Book("Chehov", "chehiy");
+        bookTolstoy = new Book("Tolstoy", "tolstolobik");
+        bookTolstoy2 = new Book("Tolstoy", "karas");
+        library.addBook(bookPuskin);
+        library.addBook(bookPuskin2);
+        library.addBook(bookPuskin3);
+        library.addBook(bookChehov);
+        library.addBook(bookTolstoy);
+        library.addBook(bookTolstoy2);
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java:64]: make it fabric method

File: homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+public class LibraryTest {
+
+    Library library;
+    Book bookPuskin;
+    Book bookPuskin2;
+    Book bookPuskin3;
+    Book bookChehov;
+    Book bookTolstoy;
+    Book bookTolstoy2;
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java:19]: This shouldn't bee here. that is test data should be provided by fabric

File: homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Test
+    public void addBook() {
+        library.addBook(bookChehov);
+        assertEquals(3,library.authorToBookNamesMap.size());
+    }
+
+    @Test
+    public void removeBook() {
+        assertTrue(library.removeBook(bookChehov));
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java:35]: Make sure that book is no longer present in library

File: homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test
+    public void findByAuthor() {
+        ArrayList<Book> extend= new ArrayList<>();
+        extend.add(bookPuskin);
+        extend.add(bookPuskin2);
+        extend.add(bookPuskin3);
```
Comment (max-java) [inline homework/PavelBo/collections/library/src/test/java/com/tutrit/java/quickstart/LibraryTest.java:43]: You can create fabric method `getPushkinBooks()`
```
* {
return List.of(
    new Book("Pushkin", "pushka"), 
    new Book("Pushkin", "duska"), 
    new Book("Pushkin", "starushka"));
}
```

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

**Pull request closed without merging**